File: Programmers/PCCP 모의고사 기출문제/2.java
import java.lang.Math;
import java.util.*;

class Solution {
    int[] result;
    boolean[][] visited;

    public int solution(int[][] land) {
        int answer = 0;
        int R=land.length;
        int C=land[0].length;
        result = new int[C];
        visited=new boolean[R][C];
        int num=1;

        for(int i=0;i<R;i++){
            for(int j=0;j<C;j++){
                if(land[i][j]==1 && !visited[i][j]){
                    bfs(i,j,land,num);
                    num+=1;
                }
            }
        }

        for(int i=0;i<C;i++){
            answer=Math.max(answer,result[i]);
        }

        return answer;
    }

    void bfs(int r, int c, int[][] land,int num){
        Queue<Node> que=new LinkedList<>();
        int R=land.length;
        int C=land[0].length;
        int[][] dirs={{0,1},{0,-1},{1,0},{-1,0}};
        visited[r][c]=true;
        que.offer(new Node(r,c));
        int cnt=1;

        int minC=C;
        int maxC=0;
        while(!que.isEmpty()){
            Node cur=que.poll();
            minC=Math.min(minC,cur.c);
            maxC=Math.max(maxC,cur.c);
            for(int i=0;i<4;i++){
                int nR=cur.r+dirs[i][0];
                int nC=cur.c+dirs[i][1];
                if(0>nR || nR>=R || 0>nC || nC>=C || visited[nR][nC] || land[nR][nC]==0)
                    continue;
                visited[nR][nC]=true;
                que.offer(new Node(nR,nC));
                cnt+=1;
            }

        }
        for(int i=minC;i<=maxC;i++){
            result[i] += cnt;
        }

        return;
    }
}
class Node{
    int r;
    int c;

    public Node(int r, int c){
        this.r=r;
        this.c=c;
    }
}

File: Programmers/PCCP 모의고사 기출문제/2.py
from collections import deque

dirs=[[0,1],[0,-1],[1,0],[-1,0]]

def OOB(r,c):
    return not (0<=r<N and 0<=c<M)

def bfs(i,j,land):
    que=deque()
    visited[i][j]=True
    size=1
    col=set()
    col.add(j)
    que.append([i,j])
    while que:
        cc,cr=que.popleft()
        for a,b in dirs:
            nr=cc+a
            nc=cr+b
            if not OOB(nr,nc) and land[nr][nc]==1 and not visited[nr][nc]:
                visited[nr][nc]=True
                size+=1
                col.add(nc)
                que.append([nr,nc])
    for c in col:
        cnt[c]+=size
    
def solution(land):
    global N,M,visited,cnt
    
    answer = 0
    N=len(land)
    M=len(land[0])
    cnt=[0 for _ in range(M)]
    visited=[[0 for _ in range(M)] for _ in range(N)]
    
    for i in range(N):
        for j in range(M):
            if land[i][j]==1 and not visited[i][j]:
                bfs(i,j,land)
    answer=max(cnt)
                    
    return answer
